risc0/circuit/rv32im/src/lib.rs

```diff
@@ -54,3 +57,35 @@ impl TapsProvider for CircuitImpl {
 }
 
 impl CircuitCoreDef<BabyBear> for CircuitImpl {}
+
+/// Fetch a control ID with the given hash, by name, and cycle limit as a power of two (po2) from
+/// the precomputed table. If the hash function is not precomputed, or the po2 is out of range,
+/// this function will return `None`.
+///
+/// Supported values for hash_name are "sha-256", "poseidon2", and "blake2b".
+#[inline]
+pub fn control_id(hash_name: impl AsRef<str>, po2: usize) -> Option<Digest> {
+    if !(risc0_zkp::MIN_CYCLES_PO2..=risc0_zkp::MAX_CYCLES_PO2).contains(&po2) {
+        return None;
+    }
+    let idx = po2 - risc0_zkp::MIN_CYCLES_PO2;
+    match hash_name.as_ref() {
+        "sha-256" => Some(SHA256_CONTROL_IDS[idx]),
+        "poseidon2" => Some(POSEIDON2_CONTROL_IDS[idx]),
+        "blake2b" => Some(BLAKE2B_CONTROL_IDS[idx]),
+        _ => None,
```

Comment:
nit: It'd be nice to somehow have a single list of strings for all hashes supported, perhaps even defined as an enum. For example we have this also: https://github.com/risc0/risc0/blob/main/risc0/zkp/src/core/hash/mod.rs#L91

Reply:
I went ahead and wrote the `enum` part of that while waiting for some builds. Using `AsRef<str>`, this can be passed into a lot of the places where we use strings right now, and possible used for the match arms as well. Actually using it throughout the codebase will touch quite a lot of places, and it's getting late, so I'll open a new PR with that tomorrow.
```patch
index 08b2ea85..b885bfcc 100644
--- a/risc0/zkp/src/core/hash/mod.rs
+++ b/risc0/zkp/src/core/hash/mod.rs
@@ -21,6 +21,7 @@ pub mod poseidon_254;
 pub mod sha;

 use alloc::{boxed::Box, rc::Rc, string::String};
+use core::str::FromStr;

 use risc0_core::field::{baby_bear::BabyBear, Field};

@@ -86,6 +87,51 @@ impl<F: Field> Clone for HashSuite<F> {
     }
 }

+/// Identifiers for hash function used in the RISC Zero codebase.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum HashId {
+    /// Identifier for the SHA-256 hash function.
+    Sha256,
+    /// Identifier for RISC Zero implementation of Poseidon2 over BabyBear.
+    Poseidon2,
+    /// Identifier for the Blak2b hash function.
+    Blake2b,
+    /// Identifier for RISC Zero implementation of Poseidon over the BN254 scalar field.
+    PoseidonBn254,
+}
+
+impl HashId {
+    /// Convert to the string representation for the identifier.
+    pub fn as_str(&self) -> &str {
+        match self {
+            Self::Sha256 => "sha-256",
+            Self::Poseidon2 => "poseidon2",
+            Self::Blake2b => "blake2b",
+            Self::PoseidonBn254 => "poseidon_254",
+        }
+    }
+}
+
+impl AsRef<str> for HashId {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl FromStr for HashId {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<HashId, Self::Err> {
+        match s {
+            "sha-256" => Ok(Self::Sha256),
+            "poseidon2" => Ok(Self::Poseidon2),
+            "blake2b" => Ok(Self::Blake2b),
+            "poseidon_254" => Ok(Self::PoseidonBn254),
+            _ => Err(()),
+        }
+    }
+}
+
 /// Construct a supported hash function given its name. Returns None is the name does not
 /// correspond to a supported hash function.
 pub fn hash_suite_from_name(name: impl AsRef<str>) -> Option<HashSuite<BabyBear>> {
 ```

Reply:
I also recommend an enum for the list of hash algorithms, which would have some benefits:
- `control_ids` no longer requires static or allocated strings
- `SegmentReceiptVerifierParameters::from_max_po2` no longer tracking this same list of algorithm names

Thoughts on your current enum (in no particular order):
- It's not being treated as a numerical ID, so perhaps a better name is `HashAlg`
- From a semantics perspective, it should not implement `AsRef<str>` since it is itself not a string buffer
- `name` might be better than `as_str`
- `FromStr::Err` should be a real error type, not `()`
- I recommend making `control_id` and `control_ids` be methods on it
- Methods on small `Copy` types that don't borrow should generally take `self` instead of `&self`

Reply:
Thanks, those are some helpful comments. I'll take them into account for the follow-up PR. As for constraints, we are not going to change interfaces covered by 1.0 semver, so the solution can't be perfect.